File: ios/AIRFacebook/Functions/ShowGameRequestDialogFunction.m
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
#import "ShowGameRequestDialogFunction.h"
#import <Foundation/Foundation.h>
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import "AIRFacebook.h"
#import "AIRFacebookShareType.h"
#import "ShareContentDelegate.h"

FREObject fb_showGameRequestDialog( FREContext context, void *functionData, uint32_t argc, FREObject *argv ) {
    NSString* message = [MPFREObjectUtils getNSString:argv[0]];
    NSString* actionType = (argv[1] == nil ) ? nil : [MPFREObjectUtils getNSString:argv[1]];
    NSString* title = (argv[2] == nil ) ? nil : [MPFREObjectUtils getNSString:argv[2]];
    NSString* objectID = (argv[3] == nil ) ? nil : [MPFREObjectUtils getNSString:argv[3]];
    NSString* friendsFilter = (argv[4] == nil ) ? nil : [MPFREObjectUtils getNSString:argv[4]];
    NSString* data = (argv[5] == nil ) ? nil : [MPFREObjectUtils getNSString:argv[5]];
    NSArray* suggestedFriends = (argv[6] == nil ) ? nil : [MPFREObjectUtils getNSArray:argv[6]];
    NSArray* recipients = (argv[7] == nil ) ? nil : [MPFREObjectUtils getNSArray:argv[7]];
    int listenerID = [MPFREObjectUtils getInt:argv[8]];

    NSMutableDictionary* params = [[NSMutableDictionary alloc] init];
    params[@"type"] = [AIRFacebookShareType toString:AIRFBShareType_GAME_REQUEST];
    params[@"message"] = message;
    params[@"listenerID"] = @(listenerID);
    if( actionType ) {
        params[@"actionType"] = actionType;
    }
    if( title ) {
        params[@"title"] = title;
    }
    if( objectID ) {
        params[@"objectID"] = objectID;
    }
    if( friendsFilter ) {
        params[@"friendsFilter"] = friendsFilter;
    }
    if( data ) {
        params[@"data"] = data;
    }
    if( suggestedFriends ) {
        params[@"suggestedFriends"] = suggestedFriends;
    }
    if( recipients ) {
        params[@"recipients"] = recipients;
    }

    /* Call the delegate */
    [[ShareContentDelegate sharedInstance] shareWithParameters:params];

    return nil;
}

File: android/src/com/marpies/ane/facebook/functions/ShowGameRequestDialogFunction.java
/*
 * Copyright (c) 2018 Marcel Piestansky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.facebook.functions;

import android.os.Bundle;
import com.adobe.fre.FREArray;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.facebook.share.model.GameRequestContent;
import com.marpies.ane.facebook.ShareActivity;
import com.marpies.ane.facebook.data.AIRFacebookShareType;
import com.marpies.ane.facebook.utils.AIR;
import com.marpies.ane.facebook.utils.FREObjectUtils;

import java.util.ArrayList;
import java.util.List;

public class ShowGameRequestDialogFunction extends BaseFunction {

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		String message = FREObjectUtils.getStringFromFREObject( args[0] );
		String actionType = (args[1] == null ) ? null : FREObjectUtils.getStringFromFREObject( args[1] );
		String title = (args[2] == null ) ? null : FREObjectUtils.getStringFromFREObject( args[2] );
		String objectID = (args[3] == null ) ? null : FREObjectUtils.getStringFromFREObject( args[3] );
		String friendsFilter = (args[4] == null ) ? null : FREObjectUtils.getStringFromFREObject( args[4] );
		String data = (args[5] == null ) ? null : FREObjectUtils.getStringFromFREObject( args[5] );
		ArrayList<String> suggestedFriends = (args[6] == null ) ? null : (ArrayList<String>)FREObjectUtils.getListOfStringFromFREArray( (FREArray)args[6] );
		ArrayList<String> recipients = (args[7] == null ) ? null : (ArrayList<String>)FREObjectUtils.getListOfStringFromFREArray( (FREArray)args[7] );
		mListenerID = FREObjectUtils.getIntFromFREObject( args[8] );

		Bundle extras = new Bundle();
		extras.putString( ShareActivity.extraPrefix + ".type", AIRFacebookShareType.GAME_REQUEST.toString() );
		extras.putString( ShareActivity.extraPrefix + ".message", message );
		extras.putInt( ShareActivity.extraPrefix + ".listenerID", mListenerID );
		if( actionType != null && !actionType.equals( "NONE" ) ) {
			extras.putString( ShareActivity.extraPrefix + ".actionType", actionType );
		}
		if( title != null ) {
			extras.putString( ShareActivity.extraPrefix + ".title", title );
		}
		if( objectID != null ) {
			extras.putString( ShareActivity.extraPrefix + ".objectID", objectID );
		}
		if( friendsFilter != null ) {
			if( friendsFilter.equals( "appUsers" ) ) {
				extras.putString( ShareActivity.extraPrefix + ".friendsFilter", GameRequestContent.Filters.APP_USERS.toString() );
			} else {
				extras.putString( ShareActivity.extraPrefix + ".friendsFilter", GameRequestContent.Filters.APP_NON_USERS.toString() );
			}
		}
		if( data != null ) {
			extras.putString( ShareActivity.extraPrefix + ".data", data );
		}
		if( suggestedFriends != null ) {
			extras.putStringArrayList( ShareActivity.extraPrefix + ".suggestedFriends", suggestedFriends );
		}
		if( recipients != null ) {
			extras.putStringArrayList( ShareActivity.extraPrefix + ".recipients", recipients );
		}

		AIR.startActivity( ShareActivity.class, extras );

		return null;
	}

}
